File: python/graphs/0841-keys-and-rooms.py
from collections import deque
from typing import List


def canVisitAllRooms(rooms: List[List[int]]) -> bool:
    keys = set({0})
    visited = set()
    
    queue = deque([0])
    
    while queue:
        cur = queue.popleft()
        cur_keys = rooms[cur]
        keys.update(cur_keys)
        
        if cur in keys:
            visited.add(cur)
            
        for key in cur_keys:
            if key not in visited:
                queue.append(key)

    return len(rooms) == len(visited)

File: cpp/graphs/0841-keys-and-rooms.cpp
#include <queue>
#include <set>
#include <vector>

using namespace std;


bool canVisitAllRooms(vector<vector<int>>& rooms) {
    set<int> keys;
    set<int> visited;
    
    queue<int> q;
    
    keys.insert(0);
    q.push(0);
    
    while(!q.empty()) {
        int cur = q.front();
        q.pop();
        
        vector<int> k = rooms[cur];
        for (int key : k) {
            keys.insert(key);
        }
        
        if (keys.find(cur) != keys.end()) {
            visited.insert(cur);
        }
        
        for (auto key : rooms[cur]) {
            if (visited.find(key) == visited.end()) {
                q.push(key);
            }
        }
    }
    
    return rooms.size() == visited.size();
}